RealmSwift/Tests/RealmCollectionTypeTests.swift

```diff
@@ -720,6 +720,127 @@ class RealmCollectionTypeTests: TestCase {
+        token0.invalidate()
+    }
+
+    func testObservePartialKeyPathNoChange() {
+        var ex = expectation(description: "initial notification")
+        let token0 = collection.observe(keyPaths: [\CTTNullableStringObjectWithLink.stringCol]) { (changes: RealmCollectionChange) in
+            switch changes {
+            case .initial(let collection):
+                XCTAssertEqual(collection.count, 2)
+            case .update:
+                XCTFail("update not expected")
+            case .error:
+                XCTFail("error not expected")
+            }
+            ex.fulfill()
+        }
+        waitForExpectations(timeout: 0.2, handler: nil)
+
+        // Expect no notification for `stringCol` key path because only `linkCol.id` will be modified.
+        ex = self.expectation(description: "NO change notification")
+        ex.isInverted = true // Inverted expectation causes failure if fulfilled.
+        dispatchSyncNewThread {
+            let realm = self.realmWithTestPath()
+            realm.beginWrite()
+            let obj = realm.objects(CTTNullableStringObjectWithLink.self).first!
+            obj.linkCol!.id = 2
+            try! realm.commitWrite()
+        }
+        waitForExpectations(timeout: 0.1, handler: nil)
+        token0.invalidate()
+    }
+
+    func testObservePartialKeyPathWithLink() {
+        var ex = expectation(description: "initial notification")
+        let token = collection.observe(keyPaths: [\CTTNullableStringObjectWithLink.linkCol?.id]) { (changes: RealmCollectionChange) in
+            switch changes {
+            case .initial(let collection):
+                XCTAssertEqual(collection.count, 2)
+            case .update(_, let deletions, let insertions, let modifications):
+                XCTAssertEqual(deletions, [])
+                XCTAssertEqual(insertions, [])
+                // The reason two column changes are expected here is because the
+                // single CTTLinkTarget object that is modified is linked to two origin objects.
+                // The 0, 1 index refers to the origin objects.
+                XCTAssertEqual(modifications, [0, 1])
+            case .error:
+                XCTFail("error not expected")
+            }
+            ex.fulfill()
+        }
+        waitForExpectations(timeout: 0.2, handler: nil)
+
+        // Only expect a change notification for `linkCol.id` keypath.
+        ex = self.expectation(description: "change notification")
+        dispatchSyncNewThread {
+            let realm = self.realmWithTestPath()
+            realm.beginWrite()
+            let obj = realm.objects(CTTNullableStringObjectWithLink.self).first!
+            obj.linkCol!.id = 2
+            try! realm.commitWrite()
+        }
+        waitForExpectations(timeout: 0.2, handler: nil)
+        token.invalidate()
+    }
+
+    func testObservePartialKeyPathWithLinkNoChangeList() {
+        var ex = expectation(description: "initial notification")
```

Comment:
This test by design should only hit ex.fulfill() once. The second expectation will fail if 826 is hit.

But fwiw, the other tests where ex.fulfill() is hit twice could do this, I suppose. See `testObservePartialKeyPathWithLink` for example.

Instead of using `ex.expectedFulfillmentCount = 2` though, there are two different expectations. And the second expectation would fail by virtue of not hitting `ex.fulfill()`. Do you think that it's better to use a single expectation with `expectedFulfillmentCount = 2`.

Reply:
you could just use `expectedFulfillmentCount = 2` for `testObservePartialKeyPathWithLink `